### PR TITLE
NPU Adaption for Sanna

### PR DESCRIPTION
**Pull request by leisuzz**
# What does this PR do?

Adaption for NPU training dreambooth lora for Sanna

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by leisuzz** [comment]
@sayakpaul Please take a look at this PR for making Sanna suitable for NPU training. Thank you so much!

File: examples/dreambooth/train_dreambooth_lora_sana.py [hunk 1]
```diff
@@ -979,10 +982,10 @@ def main(args):
         )
 
     # VAE should always be kept in fp32 for SANA (?)
-    vae.to(dtype=torch.float32)
+    vae.to(accelerator.device, dtype=torch.float32)
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_sana.py:985]: This is not needed. As we conditionally put the VAE on and off the accelerator device.

File: examples/dreambooth/train_dreambooth_lora_sana.py [hunk 2]
```diff
@@ -979,10 +982,10 @@ def main(args):
...
     transformer.to(accelerator.device, dtype=weight_dtype)
     # because Gemma2 is particularly suited for bfloat16.
-    text_encoder.to(dtype=torch.bfloat16)
+    text_encoder.to(accelerator.device, dtype=torch.bfloat16)
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_sana.py:988]: Same as https://github.com/huggingface/diffusers/pull/10409/files#r1899261106

**Review by sayakpaul — COMMENTED** [review]
Training related changes look straightforward to me. Could we also add a note about this support in the README? 

I will leave the `models` related changes for @yiyixuxu to review.

**Comment by leisuzz** [comment]
@sayakpaul Thanks for your help! You can definitely add this support in the README! By the way, I think once we've done the sd3 lora function, we should change the accelerate loading and saving functions at least for training scripts for dreambooth Lora

**Comment by sayakpaul** [comment]
> You can definitely add this support in the README!

Oh, I was wondering if you could just add a note about the NPU support in the [README](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/README_sana.md) directly. 

> By the way, I think once we've done the sd3 lora function, we should change the accelerate loading and saving functions at least for training scripts for dreambooth Lora

This is not relevant for this PR, so, we can ignore.

**Comment by leisuzz** [comment]
@sayakpaul I'm not sure what is the process, as both flux and sd don't have these notes. By the way, the npu training should be automatically proceed if the npu is available.

**Comment by sayakpaul** [comment]
Okay we can leave it out of this PR then and open a future PR to add that note.

**Comment by leisuzz** [comment]
Sure, thanks for your help!

**Review by sayakpaul — APPROVED** [review]
Changes in the training script looks good to me. Off to @yiyixuxu for the other changes.

**Comment by leisuzz** [comment]
@yiyixuxu Please take a look. Thank you!

File: src/diffusers/models/transformers/sana_transformer.py [hunk 3]
```diff
@@ -119,6 +120,12 @@ def __init__(
         # 2. Cross Attention
         if cross_attention_dim is not None:
             self.norm2 = nn.LayerNorm(dim, elementwise_affine=norm_elementwise_affine, eps=norm_eps)
+
+            if is_torch_npu_available():
```
Comment (yiyixuxu) [inline src/diffusers/models/transformers/sana_transformer.py:124]: same comment as in the other PR - let's not update default attn processor logic for now
we can manually set it

Comment (leisuzz) [reply src/diffusers/models/transformers/sana_transformer.py]: > same comment as in the other PR - let's not update default attn processor logic for now we can manually set it

I've updated the new one, please take a look. This can just use set up NPU FA directly

Comment (leisuzz) [reply src/diffusers/models/transformers/sana_transformer.py]: @yiyixuxu It still needs to modify the sanna_transformer file, so I think to check in the __init__ it;s the best option now

**Comment by leisuzz** [comment]
@yiyixuxu Please take a look at this modification, if this is fine, I will update the FLUX one as well once this has been merged. Thank you!

File: src/diffusers/models/attention_processor.py [hunk 4]
```diff
@@ -294,6 +294,10 @@ def __init__(
             processor = (
                 AttnProcessor2_0() if hasattr(F, "scaled_dot_product_attention") and self.scale_qk else AttnProcessor()
             )
+
```
Comment (yiyixuxu) [inline src/diffusers/models/attention_processor.py]: umm I don't think we should change the default attention processor here 
let's keep this logic in SANA:)

**Comment by leisuzz** [comment]
@yiyixuxu I've changed the logic back in SANA, thanks for your help!

File: src/diffusers/models/transformers/sana_transformer.py [hunk 5]
```diff
@@ -119,6 +120,13 @@ def __init__(
         # 2. Cross Attention
         if cross_attention_dim is not None:
             self.norm2 = nn.LayerNorm(dim, elementwise_affine=norm_elementwise_affine, eps=norm_eps)
+
```
Comment (yiyixuxu) [inline src/diffusers/models/transformers/sana_transformer.py:123]: @lawrence-cj let me know if it's ok with you to default to NPU attention when it's available:)

**Review by yiyixuxu — APPROVED** [review]
thanks! 
let's wait for SANA author to see if they like the default processor change, ok to me otherwise

**Comment by yiyixuxu** [comment]
actually, I think it's not so great if we only have this behavior for Sana: automatically use NPU when it is available 

maybe let's not do this unless we want to change default for everything? you can still explicitly set NPU attention processor, no?

**Comment by leisuzz** [comment]
Hi @yiyixuxu, because the default attention processor is AttnProcessor2_0, so either to change it inside model or change the set processor in the attention_processor.py

File: src/diffusers/models/transformers/sana_transformer.py (hunk 5 above)
Comment (lawrence-cj) [reply src/diffusers/models/transformers/sana_transformer.py:123]: Oh, I'm not familiar with NPU training and inference. Is this NPU device very popular in diffusers community?

**Comment by leisuzz** [comment]
Hi @yiyixuxu, is this ready to merge? Thanks

**Comment by yiyixuxu** [comment]
@leisuzz 
user can manually change it after they create the model with `set_attn_processor()` 
can you remove the change in sana? we can merge after that! :) (sorry about the back and forth)

**Comment by leisuzz** [comment]
@yiyixuxu Please take  a look at this modification and let me know if it is ok. Thanks

**Comment by leisuzz** [comment]
@yiyixuxu In fact, NPU has to use the AttnProcessorNPU, otherwise the loss will be NaN in Sana.
